### PR TITLE
PHOENIX-7004 : Validate the result returned by phoenix query is within scan range, if provided

### PR DESCRIPTION
**Pull request by Divneet18**
Jira: PHOENIX-7004

Added a check to make sure query is within start and end bounds

### CODE REVIEW

File: phoenix-core/src/main/java/org/apache/phoenix/exception/QueryOutOfScanRangeException.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.apache.phoenix.exception;
```
Comment (virajjasani) [inline phoenix-core/src/main/java/org/apache/phoenix/exception/QueryOutOfScanRangeException.java:1]: please add apache header info

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java [hunk 2]
```diff
@@ -186,20 +188,45 @@ public Tuple next() throws SQLException {
                 if (lastTuple != null) {
                     ImmutableBytesWritable ptr = new ImmutableBytesWritable();
                     lastTuple.getKey(ptr);
+                    if (scan.getStartRow() != null
+                            && Bytes.compareTo(scan.getStartRow(), HConstants.EMPTY_START_ROW) != 0) {
+                        if (scan.includeStartRow()) {
+                            if (Bytes.compareTo(scan.getStartRow(), ptr.get()) > 0) {
+                                throw new QueryOutOfScanRangeException("Query returned is out of scan range");
```
Comment (virajjasani) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:195]: nit: `\s"out of scan range"\"out of scan start rowkey range"`

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java [hunk 3]
```diff
@@ -186,20 +188,45 @@ public Tuple next() throws SQLException {
...
+                            }
+                        }
+                    }
+
+                    if (scan.getStopRow() != null
+                            && Bytes.compareTo(scan.getStopRow(), HConstants.EMPTY_END_ROW) != 0) {
+                        if (scan.includeStopRow()) {
+                            if (Bytes.compareTo(scan.getStopRow(), ptr.get()) < 0) {
+                                throw new QueryOutOfScanRangeException("Query returned is out of scan range");
```
Comment (virajjasani) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:208]: nit: `nit: \s"out of scan range"\"out of scan stop rowkey range"`

**Review by virajjasani — DISMISSED** [review]
left nits reg error msg, otherwise good to go

**Comment by shahrs87** [comment]
@Divneet18  There are 615 test failures.  https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1697/2/testReport/

**Comment by shahrs87** [comment]
Also is there any way to mock the result set and return the mocked value and make sure the following test scenarios.
1. start key is set but returned value is less than start key and the exception is thrown.
2. stop key is specified but returned value is greater than stop key and the exception is thrown.
@Divneet18

**Comment by virajjasani** [comment]
Test failures are relevant

File: phoenix-core/src/main/java/org/apache/phoenix/exception/QueryOutOfScanRangeException.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.phoenix.exception;
+
+import java.sql.SQLException;
+
+public class QueryOutOfScanRangeException extends SQLException {
```
Comment (shahrs87) [inline phoenix-core/src/main/java/org/apache/phoenix/exception/QueryOutOfScanRangeException.java:24]: s/QueryOutOfScanRangeException/ResultSetOutOfScanRangeException

**Comment by virajjasani** [comment]
Will need to revisit this, looks like we missed something fundamental here

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java [hunk 5]
```diff
@@ -186,20 +188,45 @@ public Tuple next() throws SQLException {
                 if (lastTuple != null) {
                     ImmutableBytesWritable ptr = new ImmutableBytesWritable();
                     lastTuple.getKey(ptr);
+                    if (scan.getStartRow() != null
+                            && Bytes.compareTo(scan.getStartRow(), HConstants.EMPTY_START_ROW) != 0) {
+                        if (scan.includeStartRow()) {
```
Comment (shahrs87) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:193]: What is the difference between `scan.getStartRow() != null` and `scan.includeStartRow()` ? 
Can we exclude one check if both are the same?

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java [hunk 6]
```diff
@@ -186,20 +188,45 @@ public Tuple next() throws SQLException {
...
+                        } else {
+                            if (Bytes.compareTo(scan.getStartRow(), ptr.get()) >= 0) {
+                                throw new QueryOutOfScanRangeException("Query returned is out of scan range");
+                            }
+                        }
+                    }
+
+                    if (scan.getStopRow() != null
+                            && Bytes.compareTo(scan.getStopRow(), HConstants.EMPTY_END_ROW) != 0) {
+                        if (scan.includeStopRow()) {
```
Comment (shahrs87) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:206]: What is the difference between `scan. getStopRow() != null` and `scan.includeStopRow()` ? 
Can we exclude one check if both are the same?

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java [hunk 7]
```diff
@@ -186,20 +188,45 @@ public Tuple next() throws SQLException {
...
                     ImmutableBytesWritable ptr = new ImmutableBytesWritable();
                     lastTuple.getKey(ptr);
+                    if (scan.getStartRow() != null
+                            && Bytes.compareTo(scan.getStartRow(), HConstants.EMPTY_START_ROW) != 0) {
+                        if (scan.includeStartRow()) {
+                            if (Bytes.compareTo(scan.getStartRow(), ptr.get()) > 0) {
+                                throw new QueryOutOfScanRangeException("Query returned is out of scan range");
+                            }
+                        } else {
+                            if (Bytes.compareTo(scan.getStartRow(), ptr.get()) >= 0) {
```
Comment (ranganathg) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java:198]: why is equality check required for start and stop both? are both the end exclusive?

**Review by dbwong — COMMENTED** [review]
In general where are the test cases that make sure this exception is thrown when necessary?  Especially reverse scan failure type cases.  Also validating every row seems expensive and not even possible for cases where we don't project rowkeys for a server side scan right?

**Comment by virajjasani** [comment]
Now that i think about this Jira, we are adding O(1) time complexity for every single row returned so for N rows, so this can turn out to be quite expensive, you are correct @dbwong.

@Divneet18 no worries, let's think about this and see whether we want to go forward with this change because the primary responsibility for returning the right results should stay within hbase scan and filter boundaries. Even though we check scan boundaries, what about filters? We can't introduce check for every single filter. This is going to be quite expensive.

**Comment by virajjasani** [comment]
FYI, we are resuming this effort on https://github.com/apache/phoenix/pull/1724
The latest revision has all tests passing

**Comment by virajjasani** [comment]
> FYI, we are resuming this effort on #1724 The latest revision has all tests passing

sorry for the confusion, i meant to create that PR only to check how many cases need fixes at server side, so that PR was just meant to be the initial phase for correctly identifying problems, we can continue more work in this current PR.

**Pull request closed without merging**